exporter/kafkaexporter/internal/kafkaclient/franzgo.go

```diff
@@ -0,0 +1,82 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package kafkaclient // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/kafkaexporter/internal/kafkaclient"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/kafkaexporter/internal/marshaler"
+)
+
+// FranzSyncProducer is a wrapper around the franz-go client that implements
+// the Producer interface. Allowing us to use the franz-go client while
+// maintaining compatibility with the existing Kafka exporter code.
+type FranzSyncProducer[T any] struct {
+	client       *kgo.Client
+	messenger    Messenger[T]
+	metadataKeys []string
+}
+
+// NewFranzSyncProducer Franz-go producer from a kgo.Client and a Messenger.
+func NewFranzSyncProducer[T any](
+	client *kgo.Client,
+	messenger Messenger[T],
+	metadataKeys []string,
+) FranzSyncProducer[T] {
+	return FranzSyncProducer[T]{
+		client:       client,
+		messenger:    messenger,
+		metadataKeys: metadataKeys,
+	}
+}
+
+// ExportData sends a batch of messages to Kafka
+func (p FranzSyncProducer[T]) ExportData(ctx context.Context, data T) error {
+	messages, err := partitionMessages(ctx,
+		data, p.messenger, makeFranzMessages,
+	)
+	if err != nil || len(messages) == 0 {
+		return err
+	}
```

Comment:
Rather than do the partitioning here, could we move it up a level and pass a slice of topic-messages in? Then I think we can remove the generics from the producer implementations, and keep them ignorant of signals.

Reply:
Good point, let me look into that